examples/caa/caa.ipynb

```diff
@@ -0,0 +1,590 @@
+     "text": [
+      "Evaluating: 100%|██████████| 120/120 [00:44<00:00,  2.69it/s]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Unsteered model: 0.685\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "# TODO(dtch1996): current implementation is slow...\n",
+    "result = evaluate_model(model, tokenizer, test_dataset, show_progress=True)\n",
+    "print(f\"Unsteered model: {result:.3f}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Extract Steering Vectors"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Training steering vector: 100%|██████████| 1080/1080 [06:17<00:00,  2.86it/s]\n"
+     ]
+    }
+   ],
+   "source": [
+    "from steering_vectors import train_steering_vector, SteeringVector\n",
+    "\n",
+    "steering_vector: SteeringVector = train_steering_vector(\n",
+    "    model, \n",
+    "    tokenizer,\n",
+    "    train_dataset,\n",
+    "    move_to_cpu=True,\n",
+    "    # NOTE: You can specify a list[int] of desired layer indices\n",
+    "    # If layers is None, then all layers are used\n",
+    "    # Here, layer 15 is the layer where sycophancy steering worked best in the CAA paper\n",
+    "    # for both Llama-2-7b-chat and Llama-2-13b-chat. \n",
+    "    layers = [15], \n",
+    "    # NOTE: The second last token corresponds to the A/B position\n",
+    "    # which is where we believe the model makes its decision \n",
+    "    read_token_index=-2,RESULTS[f\"steer_0\"] = result\n",
+    "    show_progress=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "SteeringVector(layer_activations={15: tensor([ 0.0644, -0.0403,  0.0907,  ..., -0.1613, -0.0146,  0.0941],\n",
+      "       dtype=torch.float16)}, layer_type='decoder_block')\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(steering_vector)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's sanity check our vector by evaluating the cosine similarity with the ground truth sycophancy vector."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Cosine similarity: 0.995\n"
+     ]
+    }
+   ],
+   "source": [
+    "original_steering_vector = torch.load(f'vec_layer_15_Llama-2-{model_size}-chat-hf.pt')\n",
+    "our_steering_vector = steering_vector.layer_activations[15]\n",
+    "\n",
+    "def calculate_cosine_similarity(a: torch.Tensor, b: torch.Tensor) -> float:\n",
```

Comment:
nit: torch already has a built-in cosine similarity function: https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html

Reply:
fixed